base_layer/wallet/src/transaction_service/protocols/transaction_coinbase_monitoring_protocol.rs

```diff
@@ -154,275 +150,443 @@ where TBackend: TransactionBackend + 'static
             loop {
                 futures::select! {
-                    base_node_response = base_node_response_receiver.select_next_some() => {
-                        match self
-                        .handle_base_node_response(completed_tx.tx_id, base_node_response)
-                        .await? {
-                            BaseNodeResponseType::ChainMetadata(result) =>
-                                chain_metadata_response_received = Some(result),
-                            BaseNodeResponseType::FetchUtxo(result) => {
-                                if result {
-                                    // Tx is mined!
-                                    return Ok(self.id);
-                                }
-                                fetch_utxo_response_received = true;
+                    new_base_node = base_node_update_receiver.select_next_some() => {
+                        match new_base_node {
+                            Ok(bn) => {
+                                self.base_node_public_key = bn;
+                                 info!(
+                                    target: LOG_TARGET,
+                                    "Coinbase Monitoring protocol (TxId: {}) Base Node Public key updated to {:?}",
+                                    self.tx_id,
+                                    self.base_node_public_key
+                                );
+                                self.first_rejection = false;
+                                continue;
                             },
-                            _ => (),
+                            Err(e) => {
+                                error!(
+                                    target: LOG_TARGET,
+                                    "Coinbase Monitoring protocol (TxId: {}) event 'base_node_update' triggered with \
+                                     error: {:?}",
+                                    self.tx_id,
+                                    e,
+                                );
+                            }
                         }
-
+                    }
+                    result = self.query_coinbase_transaction(
+                        signature.clone(), completed_tx.clone(), &mut client
+                    ).fuse() => {
+                        let (coinbase_kernel_found, metadata) = match result {
+                            Ok(r) => r,
+                            _ => (false, None),
+                        };
+                        if coinbase_kernel_found {
+                            // We are done!
+                            info!(
+                                target: LOG_TARGET,
+                                "Coinbase monitoring protocol for transaction (TxId: {}) completed successfully.",
+                                self.tx_id,
+                            );
+                            return Ok(self.tx_id);
+                        }
+                        match metadata {
```

Comment:
Style Nit: Select! statements are hard to read and the automated formatting doesn't work in them. Generally I feel its best practice to keep as much code out of them as possible. Kind of feel that this code could easily have been inside the `query_coinbase_transaction(...)` function and kept this hard to read loop much cleaner and focused on the looping/retry aspect of this section of the code.

Reply:
Code moved as suggested